ore: simplify retry API

benesch opened the pull request:
Ignore the first commit. That's in #6108.

----

The retry API in the ore crate has grown organically over time,
resulting in IMO a fairly confusing, non-orthogonal API consisting of
a RetryBuilder object, a Retry object, and a retry_for helper method.

This commit simplifies the API, unifying RetryBuilder, Retry, and
retry_for into one object named Retry. This object has a default
constructor which picks sensible retry defaults for fallible network
operations, but exposes builder methods in the case that any of the
parameter methods need tweaking.

The configuration options are whittled down to:

  * initial_backoff
  * max_backoff
  * max_tries

Exposing only these three knobs provides a good deal of flexibility
while also making it clear that Retry does not attempt to limit how long
the actual fallible operation takes. The old API

    retry_for(Duration::from_secs(60), |_| /* op */)

implied that the function would time out after 60s, when in fact the
duration only controlled the maximum permitted cumulative sleep; the
time taken by the actual fallible operation was not taken into account.

As an added bonus, the new code comes with usage examples and tests.

I've ported all callers of the old API to the new API. In most cases
I've adjusted callers that specified a custom `retry_for` duration to
just use the new default parameters, since the various hardcoded values
were not (AFAIK) chosen with a great deal of thought.  But to be safe,
anywhere where the default would result in retrying for *less* time,
I've increased the `max_tries` parameter so that the new code will retry
for at least as long as the old code would have.

For additional context, see: https://github.com/MaterializeInc/materialize/pull/5811#issuecomment-781737833

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/materializeinc/materialize/6109)
<!-- Reviewable:end -->

quodlibetor reviewed (commented):
I think that this API is much nicer from code, but it does make the sql test timeouts more opaque. How would you feel about adding an alternate constructor that allows setting a retry timeout and just initializes all the values in the `Retry` to end up at approximately that timeout?

I'm imagining something like:

```rust
/// Retry for at least the desired duration
///
/// May retry for up to twice the `desired` duration, if it is set to
/// just slightly more than an even power-of-two of the base
fn for_at_least(desired: Duration) -> Retry {
    let initial_backoff = Duration::from_millis(125);
    let mut timeout = initial_backoff.clone();
    let mut total = initial_backoff.clone();
    let mut retries = 0;
    while total < desired {
        retries += 1;
        timeout = min(timeout * 2, max_backoff);
        total += timeout;
    }
    Retry {
        initial_backoff,
        max_backoff,
        retries,
    }
}
```

mjibson reviewed (approved):
LGTM. I am however dubious about the need to pass `state` as an argument to the retry closure. I don't see it used in anything besides tests and the contrived examples. Apparently we were already doing this, but if we're rethinking the whole API it's maybe a reasonable question: do we need to keep it?

benesch commented on src/ore/src/retry.rs:
```diff
@@ -3,140 +3,270 @@
+    ///
+    /// The initial backoff is the amount of time to wait if the first try
+    /// fails.
+    pub fn initial_backoff(mut self, initial_backoff: Duration) -> Self {
```
Good idea. Done.

benesch commented:
> Apparently we were already doing this, but if we're rethinking the whole API it's maybe a reasonable question: do we need to keep it?

Ah, yeah, it's used by the testdrive SQL retry loop, which has some complicated output printing requirements that require both `state.i` and `state.next_backoff`.

benesch commented:
> I think that this API is much nicer from code, but it does make the sql test timeouts more opaque. How would you feel about adding an alternate constructor that allows setting a retry timeout and just initializes all the values in the `Retry` to end up at approximately that timeout?
> 
> I'm imagining something like:
> 
> ```rust
> /// Retry for at least the desired duration
> ///
> /// May retry for up to twice the `desired` duration, if it is set to
> /// just slightly more than an even power-of-two of the base
> fn for_at_least(desired: Duration) -> Retry {
>     let initial_backoff = Duration::from_millis(125);
>     let mut timeout = initial_backoff.clone();
>     let mut total = initial_backoff.clone();
>     let mut retries = 0;
>     while total < desired {
>         retries += 1;
>         timeout = min(timeout * 2, max_backoff);
>         total += timeout;
>     }
>     Retry {
>         initial_backoff,
>         max_backoff,
>         retries,
>     }
> }
> ```

Hmm, that's actually exactly the kind of API I was trying to avoid, because I think it's confusing for things to be in terms of the cumulative backoff rather than the elapsed time. I hear what you're saying about the SQL timeouts getting more opaque, though. So I reworked the API in the latest revision to expose a `max_duration` parameter in addition to a `max_tries` parameter. This is similar to the old `max_sleep` parameter, but tracks the total elapsed time of the operation, rather than just the cumulative backoff. I also renamed `max_backoff` to `clamp_backoff` to try to draw more of a distinction between methods that affect when the retry operation terminates vs. methods that just tune the backoff growth.

benesch commented:
I'm going to merge tomorrow afternoon unless there are objections to the new design!
